perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -4703,16 +4710,25 @@ public Group createGroupUnion(PerunSession sess, Group resultGroup, Group operan
 		// save group relation
 		groupsManagerImpl.saveGroupRelation(sess, resultGroup, operandGroup, parentFlag);
 
+		List<Member> affectedMembers = getGroupMembers(sess, operandGroup);
+
+		Map<Member, MemberGroupStatus> previousStatuses = new HashMap<>();
+		for (Member affectedMember : affectedMembers) {
+			previousStatuses.put(affectedMember, getTotalMemberGroupStatus(sess, affectedMember, resultGroup));
+		}
+
 		// do the operation logic
 		try {
-			addRelationMembers(sess, resultGroup, getGroupMembers(sess, operandGroup), operandGroup.getId());
+			addRelationMembers(sess, resultGroup, affectedMembers, operandGroup.getId());
 		} catch(AlreadyMemberException ex) {
 			throw new ConsistencyErrorException("AlreadyMemberException caused by DB inconsistency.",ex);
 		}
 
 		// calculate new member-group statuses
-		for (Member member : getGroupMembers(sess, operandGroup)) {
-			recalculateMemberGroupStatusRecursively(sess, member, operandGroup);
+		for (Member member : affectedMembers) {
+			MemberGroupStatus previousStatus = previousStatuses.get(member);
+
+			recalculateMemberGroupStatusRecursively(sess, member, operandGroup, previousStatus);
```

Comment:
```suggestion
			recalculateMemberGroupStatusRecursively(sess, member, resultGroup, previousStatus);
```
I think it should be sufficient to recalculate statuses in resultGroup, because statuses could change only there. And `previousStatuses` variable contains statuses for `resultGroup` as well.

Reply:
I don't think that would be correct, I think that there is a mistake in `previousStatuses`, where I need statuses from `operandGroup` (it should be fixed now). Although the groupStatus won't change for `operandGroup`, we need to change the status in `resultGroup` to match the status in `operandGroup` and in `recalculateMemberGroupStatusRecursively` we propage status of member to all its `resultGroups`, so we change status in `resultGroup` based on `operandGroup`.

Reply:
I think you are right. We need to call `recalculateMemberGroupStatusRecursively` on `operandGroup` because:
1. `addRelationMembers` in `createGroupUnion` inserts new indirect members to DB with default VALID status.
2. then in `recalculateMemberGroupStatusRecursively` the right status for these indirect members is set in `setIndirectGroupStatus`

But in this case, previousStatuses in `recalculateMemberGroupStatusRecursively` are wrong, because they are calculated before `setIndirectGroupStatus` and they calculate with the default VALID status inserted in `addRelationMembers`. Therefore, previousStatuses would always be VALID.

And calculating previousStatuses after `setIndirectGroupStatus` wouldn't be correct either, because they would be actually new statuses.

I am not sure about this, what do you think?

Reply:
It's again a good point, thanks. I have thought about it and I haven't come up with anything better than to send all related `MemberGroupStatuses` of the group structure to the recursive method at the beginning (as is in the new commit). I think it should work correctly now although I do not really like the solution. What do you think? Can you think of anything better?

Reply:
I can't think of anything better, but it should work correctly now. Nevertheless, I would like if some else (@balcirakpeter or @Vojtech-Sassmann) also reviewed this PR to be sure the changes won't break anything.